### ISSUE TITLE
Build failure of `examples/bluetooth/esp_hid_device` when configured to compile with performance optimization (IDFGH-9876)

### ISSUE THREAD

**Issue by finger563** [espressif/esp-idf#11197]
### Answers checklist.

- [X] I have read the documentation [ESP-IDF Programming Guide](https://docs.espressif.com/projects/esp-idf/en/latest/) and the issue is not addressed there.
- [X] I have updated my IDF branch (master or release) to the latest version and checked that the issue is present there.
- [X] I have searched the issue tracker for a similar issue and not found a similar issue.

### IDF version.

v5.1-dev-4557-gf404fe96b1

### Operating System used.

macOS

### How did you build your project?

Command line with idf.py

### If you are using Windows, please specify command line type.

None

### What is the expected behavior?

Build successfully with performance optimization configured via menuconfig

### What is the actual behavior?

![CleanShot 2023-04-14 at 09 17 15](https://user-images.githubusercontent.com/213467/232070072-ab4b5db8-f6ad-4cc7-8cdf-b81d9d3b2f3f.png)

Build failure resulting from `btc_gap_ble.c:89` when performance optimization configured via menuconfig

### Steps to reproduce.

1. Target set to ESP32
2. Set compiler optimization to `Optimize for Performance`
3. Build

### Build or installation Logs.

```plain
In file included from /Users/bob/esp/esp-idf/components/bt/host/bluedroid/btc/profile/std/gap/btc_gap_ble.c:10:
/Users/bob/esp/esp-idf/components/bt/host/bluedroid/btc/profile/std/gap/btc_gap_ble.c: In function 'btc_gap_ble_get_dev_name_callback':
/Users/bob/esp/esp-idf/components/bt/host/bluedroid/stack/include/stack/bt_types.h:35:34: error: 'strncpy' specified bound 33 equals destination size [-Werror=stringop-truncation]
   35 | #define BCM_STRNCPY_S(x1,x2,x3)  strncpy((x1),(x2),(x3))
      |                                  ^~~~~~~~~~~~~~~~~~~~~~~
/Users/bob/esp/esp-idf/components/bt/host/bluedroid/btc/profile/std/gap/btc_gap_ble.c:89:9: note: in expansion of macro 'BCM_STRNCPY_S'
   89 |         BCM_STRNCPY_S(param.get_dev_name_cmpl.name, name, BTC_MAX_LOC_BD_NAME_LEN + 1);
      |         ^~~~~~~~~~~~~
cc1: some warnings being treated as errors
[903/1189] Building C object esp-idf/bt/CMakeFiles/__idf_bt.dir/host/bluedroid/btc/profile/std/gatt/btc_gatts.c.objninja: build stopped: subcommand failed.
```

### More Information.

Happens on `esp-idf/main` (which is where the logs above come from) as well as `esp-idf/release/v5.1`, but doesn't seem to happen (in our cloud builds) on `release/v5.0`

**Comment by finger563** [comment]
It does not happen when optimization is set to `Optimize for size` - it _only_ happens when set to `Optimize for performance`

(Of course the example cannot link when set to `Debug without optimization`, but the compilation at least doesn't fail)

**Comment by xiewenxiang** [comment]
Hi @finger563 ,

Please apply this patch and test again.

[0001-component-bt-Fix-Build-failure-when-configured-to-co.zip](https://github.com/espressif/esp-idf/files/11257478/0001-component-bt-Fix-Build-failure-when-configured-to-co.zip)

**Comment by finger563** [comment]
Hi @xiewenxiang yes that's almost the same patch that I'd made locally to get around the issue, thanks. Can we expect this in a 5.1 update or 5.2?

**Comment by ftab** [comment]
This issue occurs in release/v5.0 now

**Comment by esp-cjh** [comment]
Hi, @ftab 
We have fixed this issue, but the commit hasn't been merged on release/v5.0.

**Comment by zhp0406** [comment]
Commit fixed in version 5.1: 62f5905be25e8f096032fb0f62dd768e7f0008de

**Comment by Alvin1Zhang** [comment]
Thanks for reporting, feel free to reopen.
